File: mobile/lib/protocol/auth_denied_payload.dart
import 'dart:typed_data';

import 'frame.dart';
import 'payload_codec.dart';

const int authDeniedReasonBadCredential = 0x01;
const int authDeniedReasonExpiredCode = 0x02;
const int authDeniedReasonDeviceLimit = 0x03;

class AuthDeniedPayload {
  const AuthDeniedPayload({
    required this.reason,
    required this.message,
  });

  final int reason;
  final String message;

  @override
  bool operator ==(Object other) =>
      other is AuthDeniedPayload &&
      reason == other.reason &&
      message == other.message;

  @override
  int get hashCode => Object.hash(reason, message);
}

class AuthDeniedPayloadCodec {
  static Uint8List encode(AuthDeniedPayload payload) {
    if (!_isKnownReason(payload.reason)) {
      throw const ProtocolException('Unknown AUTH_DENIED reason.');
    }

    final writer = PayloadWriter()
      ..writeUInt8(payload.reason)
      ..writeString(payload.message);
    return writer.toBytes();
  }

  static AuthDeniedPayload decode(List<int> bytes) {
    final reader = PayloadReader(bytes);
    final reason = reader.readUInt8();
    final message = reader.readString();
    reader.expectEnd();

    if (!_isKnownReason(reason)) {
      throw const ProtocolException('Unknown AUTH_DENIED reason.');
    }

    return AuthDeniedPayload(reason: reason, message: message);
  }

  static bool _isKnownReason(int reason) =>
      reason == authDeniedReasonBadCredential ||
      reason == authDeniedReasonExpiredCode ||
      reason == authDeniedReasonDeviceLimit;
}

File: desktop/Protocol/AuthDeniedPayload.cs
namespace CTRL.Desktop.Protocol;

public sealed record AuthDeniedPayload(
    byte Reason,
    string Message);

public static class AuthDeniedPayloadCodec
{
    public const byte ReasonBadCredential = 0x01;
    public const byte ReasonExpiredCode = 0x02;
    public const byte ReasonDeviceLimit = 0x03;

    public static byte[] Encode(AuthDeniedPayload payload)
    {
        if (!IsKnownReason(payload.Reason))
            throw new ProtocolException("Unknown AUTH_DENIED reason.");

        var writer = new PayloadWriter();
        writer.WriteUInt8(payload.Reason);
        writer.WriteString(payload.Message);
        return writer.ToArray();
    }

    public static AuthDeniedPayload Decode(byte[] payload)
    {
        var reader = new PayloadReader(payload);
        var reason = reader.ReadUInt8();
        var message = reader.ReadString();
        reader.ExpectEnd();

        if (!IsKnownReason(reason))
            throw new ProtocolException("Unknown AUTH_DENIED reason.");

        return new AuthDeniedPayload(reason, message);
    }

    private static bool IsKnownReason(byte reason) =>
        reason == ReasonBadCredential ||
        reason == ReasonExpiredCode ||
        reason == ReasonDeviceLimit;
}
